### PR TITLE
Bump to 4.2.1 and Add the bump version helper

### PR DESCRIPTION
**Pull request by andyduong1920**
## What happened

As the `bump version` step is kind of similar on each release. Creating a task for it.

## Insight

1. Create 1 Elixir module
2. Create 1 Elixir task
3. Run it with `mix nimble_template.bump_version xxx`

### I/ The Elixir module

1/ 
The test is passed

2/ mix help nimble_template.bump_version
image

3/ Test the task
image

4/ Invalid format

image

### II/ Github Action

1/ Invalid case

image

2/ Valid case

image

The demo PR: https://github.com/nimblehq/elixir-templates/pull/224

### CODE REVIEW

File: .github/workflows/bump_version.yml [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+      - name: Bump version
+        run: mix nimble_template.bump_version ${{ github.event.inputs.newVersion }}
+
+      - name: Create Pull Request
+        uses: peter-evans/create-pull-request@v4
+        with:
+          token: ${{ secrets.WIKI_ACTION_TOKEN }}
+          commit-message: Bump to ${{ github.event.inputs.newVersion }}
+          committer: Dev Nimble <dev@nimblehq.co>
```
Comment (olivierobert) [inline .github/workflows/bump_version.yml:76]: We should use the bot user: 
```suggestion
          committer: Nimble Bot <bot@nimblehq.co>
```

File: .github/workflows/bump_version.yml [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+          branch: chore/bump-to-${{ github.event.inputs.newVersion }}
+          delete-branch: true
+          title: [Chore] Bump to ${{ github.event.inputs.newVersion }}
```
Comment (olivierobert) [inline .github/workflows/bump_version.yml]: How about:

```suggestion
          title: [Chore] Bump version to ${{ github.event.inputs.newVersion }}
```

File: README.md [hunk 3]
```diff
@@ -71,9 +71,17 @@ Set the `HEX_API_KEY` as a Github secret (skip this step if it has been done).
 
 The release process follows the [Git flow](https://nimblehq.co/compass/development/version-control/release-management).
 
-Once a `release/<version number>` is created, to publish the new version to Hex.pm, the version number in the `mix.ex` file needs to be updated on the release branch before merging.
+1. Bump the new version on local
```
Comment (olivierobert) [inline README.md:74]: How about:

```suggestion
1. Bump the application version on your local machine:
```

File: README.md [hunk 4]
```diff
@@ -71,9 +71,17 @@ Set the `HEX_API_KEY` as a Github secret (skip this step if it has been done).
...
 
-Once the release branch is merged into the `master` branch, Github Action automatically publishes the template to [https://hex.pm/packages/nimble_template](https://hex.pm/packages/nimble_template).
+  ```bash
+  mix nimble_template.bump_version new_version
+  ```
+
+2. Open a PR to develop.
+
+3. Once that bump_version PR is merged, create the `release/<version number>` from the `develop` branch, points to the `master` branch.
```
Comment (olivierobert) [inline README.md:82]: Minor English corrections:

```suggestion
3. Once the bump_version PR is merged, create the `release/<version number>` from the `develop` branch, points to the `main` branch.
```

File: README.md [hunk 5]
```diff
@@ -71,9 +71,17 @@ Set the `HEX_API_KEY` as a Github secret (skip this step if it has been done).
...
 
-Once a `release/<version number>` is created, to publish the new version to Hex.pm, the version number in the `mix.ex` file needs to be updated on the release branch before merging.
+1. Bump the new version on local
 
-Once the release branch is merged into the `master` branch, Github Action automatically publishes the template to [https://hex.pm/packages/nimble_template](https://hex.pm/packages/nimble_template).
+  ```bash
+  mix nimble_template.bump_version new_version
+  ```
+
+2. Open a PR to develop.
```
Comment (olivierobert) [inline README.md:80]: Minor inconsistency as branches are wrapped into back-ticks below:

```suggestion
2. Open a PR to the `develop` branch.
```

File: lib/mix/tasks/nimble_template.bump_version.ex [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  @moduledoc """
+  #{@shortdoc}
+
+  - Hex package: https://hex.pm/packages/nimble_template
+  - Github: https://github.com/nimblehq/elixir-templates
+
+  # Usage
+
+  - mix help nimble_template.bump_version # Print help
+  - mix nimble_template.bump_version [new_version] # Bump the template into the [new_version].
```
Comment (olivierobert) [inline lib/mix/tasks/nimble_template.bump_version.ex:13]: How about:

```suggestion
  - mix nimble_template.bump_version [new_version] # Bump the template version to the [new_version].
```

File: lib/nimble_template/version.ex [hunk 7]
```diff
@@ -0,0 +1,31 @@
+defmodule NimbleTemplate.Version do
+  @moduledoc false
+
+  alias NimbleTemplate.Generator
+
+  def bump(new_version) do
+    current_version = Mix.Project.config()[:version]
+
+    if new_version > current_version do
```
Comment (olivierobert) [inline lib/nimble_template/version.ex:9]: Even after checking the tests, I am not sure how this comparison works 😆 How can it compare strings?

**Review by olivierobert — COMMENTED** [review]
Some early feedback for you 🤗

File: lib/nimble_template/version.ex (hunk 7 above)
Comment (andyduong1920) [reply lib/nimble_template/version.ex:9]: It seems like Elixir support compares the same string pattern, eg: with the same pattern `x.x.x` Elixir can compare `4.0.0` vs `4.0.1`

image

File: .github/workflows/bump_version.yml (hunk 1 above)
Comment (andyduong1920) [reply .github/workflows/bump_version.yml]: adjusted it in https://github.com/nimblehq/elixir-templates/pull/219/commits/8a0f97488fdf4bc0da852e8eeab73f737e256c55, we can see it on https://github.com/nimblehq/elixir-templates/pull/224

File: .github/workflows/bump_version.yml (hunk 2 above)
Comment (andyduong1920) [reply .github/workflows/bump_version.yml]: adjusted it in https://github.com/nimblehq/elixir-templates/pull/219/commits/8a0f97488fdf4bc0da852e8eeab73f737e256c55, we can see it on https://github.com/nimblehq/elixir-templates/pull/224

File: README.md (hunk 4 above)
Comment (andyduong1920) [reply README.md:82]: as we are now having the Github Action there, will update the Readme content :)

File: lib/mix/tasks/nimble_template.bump_version.ex (hunk 6 above)
Comment (andyduong1920) [reply lib/mix/tasks/nimble_template.bump_version.ex]: updated in https://github.com/nimblehq/elixir-templates/pull/219/commits/4b0f3f281db7c58f8b52e4e75f6ab469ff79e985

**Review by olivierobert — APPROVED** [review]

**Review by nvminhtue — APPROVED** [review]

**Review by Nihisil — APPROVED** [review]

**Pull request merged**